src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/TypeMapLazyDictionary.cs

```diff
@@ -65,25 +65,51 @@ private static unsafe partial void ProcessAttributes(
             delegate* unmanaged<CallbackContext*, ProcessAttributesCallbackArg*, void> newProxyTypeEntry,
             CallbackContext* context);
 
-        private static TypeName CreateNewTypeName(TypeName typeName, ReadOnlySpan<char> assemblyName)
+        private static TypeName NewTypeNameWithAssembly(TypeName typeName, ReadOnlySpan<char> assemblyName)
         {
             AssemblyNameInfo asmNameInfo = AssemblyNameInfo.Parse(assemblyName);
             TypeName parsedType = typeName.WithAssemblyName(asmNameInfo);
             Debug.Assert(parsedType.AssemblyName != null);
             return parsedType;
         }
 
-        private static ReadOnlySpan<char> ConvertUtf8ToUtf16(ReadOnlySpan<byte> utf8TypeName)
+        public ref struct Utf16SharedBuffer
         {
-            int defaultUtf8StrLen = 1024;
+            private char[]? _backingArray;
+            public Utf16SharedBuffer()
+            {
+                _backingArray = null;
+                Buffer = default;
+            }
+
+            public Utf16SharedBuffer(char[] backingBuffer, int validLength)
+            {
+                _backingArray = backingBuffer;
+                Buffer = new ReadOnlySpan<char>(backingBuffer, 0, validLength);
+            }
+
+            public ReadOnlySpan<char> Buffer { get; init; }
+
+            public void Dispose()
+            {
+                if (_backingArray != null)
+                {
+                    ArrayPool<char>.Shared.Return(_backingArray);
+                }
+            }
+        }
+
+        private static void ConvertUtf8ToUtf16(ReadOnlySpan<byte> utf8TypeName, out Utf16SharedBuffer utf16Buffer)
+        {
+            const int DefaultCharArrayLen = 1024;
             const int MaxUtf8BytesPerChar = 3;
-            int needed = utf8TypeName.Length > (defaultUtf8StrLen * MaxUtf8BytesPerChar)
+            int needed = utf8TypeName.Length > (DefaultCharArrayLen * MaxUtf8BytesPerChar)
                 ? Encoding.UTF8.GetCharCount(utf8TypeName)
-                : defaultUtf8StrLen;
+                : DefaultCharArrayLen;
```

Comment:
```suggestion
                : utf8TypeName.Length;
```

Reply:
I thought it would be better to just keep asking for the same size. The "length" here is going to be a bit odd because we're now mixing units. I think the default value is clearer.

Reply:
Would it be better to avoid going into the UTF8 encoding details and use the UTF 8 encoding method for the conservative estimate?

```
// Use quick conservative estimate for small strings
int needed = (utf8TypeName.Length < 1024) ? Encoding.UTF8.GetMaxCharCount(utf8TypeName.Length) : Encoding.UTF8.GetCharCount(utf8TypeName);
```

Reply:
> Would it be better to avoid going into the UTF8 encoding details and use the UTF 8 encoding method for the conservative estimate?
> 
> ```
> // Use quick conservative estimate for small strings
> int needed = (utf8TypeName.Length < 1024) ? Encoding.UTF8.GetMaxCharCount(utf8TypeName.Length) : Encoding.UTF8.GetCharCount(utf8TypeName);
> ```

Yes, that makes perfect sense. Thanks!